# <p style='text-align: center;'> Discretization </p>

## 2. Discretization :
- Discretization is the process of transforming continuous variables into discrete variables by creating a set of continuous 
  intervals that span the range of the variables values.
  
  
- Discretization is also called binning, where bin is an alternative name for interval.


- Discretization helps handle outliers by placing these values into the lower or highr intervals, together with the remaining 
  inlier values of the distribution. Thus, these outlier observations no longer differ from the rest of the values at the tails 
  of the distribution, as they are now all together in the same interval / bucket.
  
  
- In addition, by creating appropriate bins or intervals, discretization can help spread the values of a skewed variable across 
  a set of bins with equal number of observations.
  

In [4]:
# Importing the necessary libraries.
import pandas as pd

''' Loading "stroke prediction.csv" data into a pandas dataframe, store
    in DataFrame nams as "df" and printing out its sample.'''
df = pd.read_csv("stroke prediction.csv")

# print the top 5 records with variables.
df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0  30669    Male   3.0             0              0           No   
1  30468    Male  58.0             1              0          Yes   
2  16523  Female   8.0             0              0           No   
3  56543  Female  70.0             0              0          Yes   
4  46136    Male  14.0             0              0           No   

      work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0      children          Rural              95.12  18.0              NaN   
1       Private          Urban              87.96  39.2     never smoked   
2       Private          Urban             110.89  17.6              NaN   
3       Private          Rural              69.04  35.9  formerly smoked   
4  Never_worked          Rural             161.28  19.1              NaN   

   stroke  
0       0  
1       0  
2       0  
3       0  
4       0

In [5]:
# Get the size of data points by printing its shape.
df.shape

(43400, 12)

- df DataFrame consists of 43400 rows and 12 columns.

In [6]:
# Get the counts unique values from "stroke" variable by using value_counts().
df["stroke"].value_counts()

0    42617
1      783
Name: stroke, dtype: int64

-  Counted Patients that suffer from stroke is 783 patients and those with did not suffer from stroke is 42617.

In [8]:
# WAPP to demonstrate discretization technique.

# Create bins for the "age" attribute.
intervals = [0, 12, 19, 30, 60, 90]

# Replace the actual data values with the bins label.
categories = ["child", "teenager", "young_adult", "middle_aged", "senior_citizen"]

'''Apply discretization using intervals (bins) and the same is 
   stored in new column called “age_category”.'''

df["age_category"] = pd.cut(df["age"], bins=intervals, labels=categories)

# print the result with sample dataset.
df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0  30669    Male   3.0             0              0           No   
1  30468    Male  58.0             1              0          Yes   
2  16523  Female   8.0             0              0           No   
3  56543  Female  70.0             0              0          Yes   
4  46136    Male  14.0             0              0           No   

      work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0      children          Rural              95.12  18.0              NaN   
1       Private          Urban              87.96  39.2     never smoked   
2       Private          Urban             110.89  17.6              NaN   
3       Private          Rural              69.04  35.9  formerly smoked   
4  Never_worked          Rural             161.28  19.1              NaN   

   stroke    age_category  
0       0           child  
1       0     middle_aged  
2       0           child  
3       0  senior_citizen  
4       0        teenager

- In the above example, we have Created bins for the "age" attribute as "intervals" and Replaced the actual data values with 
  the bins label like : categories=["child", "teenager", "young_adult", "middle_aged", "senior_citizen"]. Then applied the 
  discretization using intervals (bins), labels=categories and the same is stored in new column called “age_category”.
  
  
- Here we can see that, we have created one new column as "age_category" with different categories based on the "age"  
  intervals, as foolows :
  
      age(bins)          category
      ---------          --------
      0 to 12            child
      12 to 19           teenager
      19 to 30           young_adult
      30 to 60           middle_aged
      60 to 90           senior_citizen
      
      
- Below we are displaying "age" and "age_category" variables.

In [13]:
# Display "age" and "age_category" variables.
df[["age", "age_category"]]

age    age_category
0       3.0           child
1      58.0     middle_aged
2       8.0           child
3      70.0  senior_citizen
4      14.0        teenager
...     ...             ...
43395  10.0           child
43396  56.0     middle_aged
43397  82.0  senior_citizen
43398  40.0     middle_aged
43399  82.0  senior_citizen

[43400 rows x 2 columns]

In [14]:
# Get the counts unique values from "age_category" variable by using value_counts().
df["age_category"].value_counts()

middle_aged       18653
senior_citizen    10511
young_adult        5725
child              5326
teenager           3185
Name: age_category, dtype: int64

- Counted age_categories based on there categories.

In [15]:
# Apply crosstab on "age_category" and "stroke" variables.
pd.crosstab(df.age_category, df.stroke)

stroke              0    1
age_category              
child            5325    1
teenager         3184    1
young_adult      5723    2
middle_aged     18453  200
senior_citizen   9932  579

- From the crosstab, we can see that Countes of Patients that suffer / did not suffer from stroke based on the age_category as follows :


      - 1 child is suffering from the stroke and 5325 child's are not suffering from the stroke. 
      - 1 teenager is suffering from the stroke and 3184 teenager's are not suffering from the stroke.
      - 2 young_adult's are suffering from the stroke and 5723 young_adult's are not suffering from the stroke.      
      - 200 middle_age's are suffering from the stroke and 18453 middle_age's are not suffering from the stroke.
      - 579 senior_citizen's are suffering from the stroke and 9932 senior_citizen's are not suffering from the stroke.


